### PR TITLE
m4ize SOS bindings

### PR DESCRIPTION
**Pull request by jdinan**
Generate SOS headers, profiling interfaces, and language bindings using m4.

This is a significant improvement over the C preprocessor/macro approach that we have been using.  Using m4 allows us to generate the preprocessor magic needed for the profiling interfaces.  It also allows us to generate proper header files, so that usage errors are reported to the user in terms of the function prototype, rather than an error occurring somewhere in a sequence of macros.

### CODE REVIEW

**Comment by jpdoyle** [comment]
This is definitely a better tool for the job. I'll be honest, I hadn't heard about M4 before this pull request and I wish I had.

**Comment by jdinan** [comment]
Wow, the Travis build is finally warnings-free.  :)

**Comment by jdinan** [comment]
Closes #49

**Comment by jdinan** [comment]
Closes #280

**Comment by jdinan** [comment]
Closes #279

File: mpp/shmem.h4.in [hunk 1]
```diff
@@ -0,0 +1,321 @@
...
+#define  SHMEM_MAJOR_VERSION 1
+#define _SHMEM_MAJOR_VERSION SHMEM_MAJOR_VERSION
+#define  SHMEM_MINOR_VERSION 3
+#define _SHMEM_MINOR_VERSION SHMEM_MINOR_VERSION
+#define  SHMEM_MAX_NAME_LEN  256
+#define _SHMEM_MAX_NAME_LEN  SHMEM_MAX_NAME_LEN
+#define  SHMEM_VENDOR_STRING "Sandia OpenSHMEM"
+#define _SHMEM_VENDOR_STRING SHMEM_VENDOR_STRING
+
+define(`SHFN', `shmem_')dnl
```
Comment (jdinan) [inline mpp/shmem.h4.in]: Delete this line

File: src/shmem_bind_f.m4 [hunk 2]
```diff
@@ -0,0 +1,48 @@
+dnl Copyright (c) 2016 Intel Corporation. All rights reserved.
+dnl This software is available to you under the BSD license.
+dnl
+dnl This file is part of the Sandia OpenSHMEM software package. For license
+dnl information, see the LICENSE file in the top level directory of the
+dnl distribution.
+dnl
+dnl This file contains m4 macros used to generate the SOS Fortran bindings.
+dnl
+define(`SHMEM_BIND_F_RMA', dnl args: macro_name
```
Comment (jdinan) [inline src/shmem_bind_f.m4]: Delete args comments

**Comment by jdinan** [comment]
@kseager - Cleanups from today's review are committed.  All clear from my end.

**Comment by kseager** [comment]
👍 worked locally for me 😄

**Pull request merged**